### ISSUE TITLE
PostgresProtocolPlugin not supporting Datasource Pooling

### ISSUE THREAD

**Issue by tariqdweik** [ArcadeData/arcadedb#1325]
### ArcadeDB Version:
`ArcadeDB Server v23.10.1 (build 237a95c415ca75df9f16f0852eea55a76a9e94f7/1698808521864/main)`
### OS and JDK Version:
`Windows 11 10.0 - OpenJDK 64-Bit Server VM 17.0.9 (Temurin-17.0.9+9)`

### Expected behavior
`Server should accept connections originated from connection pools and/or created from datasources`

### Actual behavior
`Server is throwing the following exception

Unexpected message type '' for message any
com.arcadedb.postgres.PostgresProtocolException: Unexpected message type '' for message any
        at com.arcadedb.postgres.PostgresNetworkExecutor.readMessage(PostgresNetworkExecutor.java:1006)
        at com.arcadedb.postgres.PostgresNetworkExecutor.run(PostgresNetworkExecutor.java:131)
`

### Steps to reproduce
`below is using dbcp2 library

import java.sql.Connection;
import java.sql.PreparedStatement;
import java.sql.ResultSet;
import java.sql.SQLException;

import org.apache.commons.dbcp2.BasicDataSource;

public class ArcadeDBPoolingTest {
	private static BasicDataSource dataSource;

	private static BasicDataSource getDataSource() {

		if (dataSource == null) {
			BasicDataSource ds = new BasicDataSource();
			ds.setUrl("jdbc:postgresql://localhost/OpenBeer");
			ds.setUsername("root");
			ds.setPassword("admin123");

			ds.setMinIdle(5);
			ds.setMaxIdle(10);
			ds.setMaxOpenPreparedStatements(100);

			dataSource = ds;
		}
		return dataSource;
	}

	public static void main(String[] args) throws SQLException {

		try (BasicDataSource dataSource = ArcadeDBPoolingTest.getDataSource();
				Connection connection = dataSource.getConnection();
				PreparedStatement pstmt = connection.prepareStatement("SELECT 1");) {
			System.out.println("The Connection Object is of Class: " + connection.getClass());
			try (ResultSet resultSet = pstmt.executeQuery();) {
				while (resultSet.next()) {
					System.out.println(resultSet.getInt(1));
				}
			} catch (Exception e) {
				connection.rollback();
				e.printStackTrace();
			}
		}
	}

}
`

`The same result is when using HikariCP and Tomcat Pooling libraries`

**Comment by lvca** [comment]
@tariqdweik what does it change when it's from a pool library? Shouldn't be the same?

**Comment by tariqdweik** [comment]
After some investigation here are my finding:

1. Postgres Plugin requires that any connection to include an SQL statement, the pooling libraries open a connection with empty statement, the workaround if to set the option of executing SQL at the time for initiating the connection or testing the connection when retrieving the connection from the pool, below is the code sample highlighting the workaround.

The workaround worked with DBCP2, below the the code:
```
import java.sql.Connection;
import java.sql.PreparedStatement;
import java.sql.ResultSet;
import java.sql.SQLException;

import org.apache.commons.dbcp2.BasicDataSource;

public class DBCPTest {
	private static BasicDataSource dataSource;

	private static BasicDataSource getDataSource() {

		if (dataSource == null) {
			BasicDataSource ds = new BasicDataSource();
			ds.setUrl("jdbc:postgresql://localhost/OpenBeer");
			ds.setUsername("root");
			ds.setPassword("admin123");

			ds.setMinIdle(5);
			ds.setMaxIdle(10);
			ds.setMaxOpenPreparedStatements(100);
			
			ds.setValidationQuery("select 1"); //<-- Workaround

			dataSource = ds;
		}
		return dataSource;
	}

	public static void main(String[] args) throws SQLException {
		try (BasicDataSource dataSource = DBCPTest.getDataSource();
				Connection connection = dataSource.getConnection();
				PreparedStatement pstmt = connection.prepareStatement("SELECT 1");) {
			System.out.println("The Connection Object is of Class: " + connection.getClass());
			try (ResultSet resultSet = pstmt.executeQuery();) {
				while (resultSet.next()) {
					System.out.println(resultSet.getInt(1));
				}
			} catch (Exception e) {
				connection.rollback();
				e.printStackTrace();
			}
		}
	}

}
```

However, the workaround didn't work in HikariCP and gives the following exception:
```
ERROR com.zaxxer.hikari.pool.HikariPool -- HikariPool-1 - Exception during pool initialization.
org.postgresql.util.PSQLException: No results were returned by the query.
	at org.postgresql.jdbc.PgConnection.execSQLQuery(PgConnection.java:576)
	at org.postgresql.jdbc.PgConnection.execSQLQuery(PgConnection.java:562)
	at org.postgresql.jdbc.PgConnection.getTransactionIsolation(PgConnection.java:1023)
	at com.zaxxer.hikari.pool.PoolBase.checkDefaultIsolation(PoolBase.java:480)
	at com.zaxxer.hikari.pool.PoolBase.checkDriverSupport(PoolBase.java:443)
```
which leads us to the second finding

2.  When using HikariCP we go the following line in the exception stack
```
	at org.postgresql.jdbc.PgConnection.getTransactionIsolation(PgConnection.java:1023)
```

PgConnection.java line 1023
```
	final ResultSet rs = execSQLQuery("SHOW TRANSACTION ISOLATION LEVEL");
```
the query ```SHOW TRANSACTION ISOLATION LEVEL``` is not implemented neither by Postgres Plugin nor the server itself (tested by running it in the console and returned error)

Also, when testing with Tomcat JDBC pooling we got the following line in the exception stack
```
	at org.postgresql.jdbc.PgConnection.getMetaData(PgConnection.java:1393)
```
and examining ```org.postgresql.jdbc.PgDatabaseMetaData``` shows how it is full of queries that are Postgres-specific

In summary (and my opinion) we have the following issues:
**1. Tolerating empty messages**: Postgres Plugin (```com.arcadedb.postgres.PostgresNetworkExecutor.readMessage```) should tolerate empty messages

**2. Insufficient JDBC Driver Implementation**: Postgres Plugin lack the full coverage of calls/queries made by Postgres JDBC driver therefore using it currently not reliable, so it is either to fully pass the unit tests of Postrgres JDBC Driver (https://github.com/pgjdbc/pgjdbc/tree/master/pgjdbc/src/test) or implement ArcadeDB's own JDBC driver

**Comment by lvca** [comment]
@tariqdweik I've just supported the command `SHOW TRANSACTION ISOLATION LEVEL` in the Postgres driver. For any other SHOW command, the database name is returned, hoping it's enough to pass the driver initialization.

Since ArcadeDB is not Postgres, we don't have all its commands and table names. Our approach is to catch and fix any issue with the most common lib/framework. We notice every driver/lib/framework runs some query at connection time and if we pass that phase they usually work.

Also, the Postgres wire protocol is not well documented. Many workflows are totally missing, like the recent FLUSH I supported today, it's not stated what it should return.

@tariqdweik Is there any chance you can check HikariCP again against the latest snapshot?

**Comment by lvca** [comment]
About the Tomcat CP, I found in the documentation a simple usage:

```java
import java.sql.Connection;
  import java.sql.ResultSet;
  import java.sql.Statement;

  import org.apache.tomcat.jdbc.pool.DataSource;
  import org.apache.tomcat.jdbc.pool.PoolProperties;

  public class SimplePOJOExample {

      public static void main(String[] args) throws Exception {
          PoolProperties p = new PoolProperties();
          p.setUrl("jdbc:mysql://localhost:3306/mysql");
          p.setDriverClassName("com.mysql.jdbc.Driver");
          p.setUsername("root");
          p.setPassword("password");
          p.setJmxEnabled(true);
          p.setTestWhileIdle(false);
          p.setTestOnBorrow(true);
          p.setValidationQuery("SELECT 1");
          p.setTestOnReturn(false);
          p.setValidationInterval(30000);
          p.setTimeBetweenEvictionRunsMillis(30000);
          p.setMaxActive(100);
          p.setInitialSize(10);
          p.setMaxWait(10000);
          p.setRemoveAbandonedTimeout(60);
          p.setMinEvictableIdleTimeMillis(30000);
          p.setMinIdle(10);
          p.setLogAbandoned(true);
          p.setRemoveAbandoned(true);
          p.setJdbcInterceptors(
            "org.apache.tomcat.jdbc.pool.interceptor.ConnectionState;"+
            "org.apache.tomcat.jdbc.pool.interceptor.StatementFinalizer");
          DataSource datasource = new DataSource();
          datasource.setPoolProperties(p);

          Connection con = null;
          try {
            con = datasource.getConnection();
            Statement st = con.createStatement();
            ResultSet rs = st.executeQuery("select * from user");
            int cnt = 1;
            while (rs.next()) {
                System.out.println((cnt++)+". Host:" +rs.getString("Host")+
                  " User:"+rs.getString("User")+" Password:"+rs.getString("Password"));
            }
            rs.close();
            st.close();
          } finally {
            if (con!=null) try {con.close();}catch (Exception ignore) {}
          }
      }

  }
```

We could easily include this in our test cases.

**Comment by lvca** [comment]
Tried to remove `p.setValidationQuery("SELECT 1");` from tomcat test and still works. I'm closing this issue. In case you find anything relevant to this issue, please comment here to reopen it.
